File: clibrary/costfunc.cpp
#include "su.h"
#include "segy.h"
#include "Complex.h"
#include "clibrary.h"
float costfunc(complex *d, complex *dc, complex *u,int nh,int nq,
float sigma,int norma)
{
      int i;
      float  r2, power,cost;

      r2=0;
      for(i=0;i<nh;i++)
         r2=r2+abs((d[i]-dc[i]))*abs((d[i]-dc[i]));
      
      power=0;
      if (norma==10) 
          for(i=0;i<nq;i++)
              power=power+log(1+real(conjg(u[i])*u[i])/sigma);            
      else if(norma==1) 
          for(i=0;i<nq;i++)
              power=power+abs(u[i]);
          
      
      cost=r2+power;

      //     write(0,*) 'Jtot,Jdata,Jmodel',costfunc,r2,power
      
      
      return(cost);
}
      

      

File: flibrary/costfunc.f
      function costfunc(d,dc,u,nx,np,sigma,norma)

      implicit none
      integer  nx, np,i, norma
      complex*16 u(*), d(*), dc(*)
      real*8 sigma,r2, costfunc, power

      r2=0
      do i=1,nx
         r2=r2+zabs((d(i)-dc(i)))**2
      enddo
      power=0.d0
      if (norma.eq.10) then
          do i=1,np
              power=power+dlog(1+dreal(dconjg(u(i))*u(i))/sigma)
          enddo
            
      elseif(norma.eq.1) then
          do i=1,np
              power=power+dcmplx(zabs(u(i)),0.d0)
          enddo
      endif
      costfunc=r2+power

c     write(0,*) 'Jtot,Jdata,Jmodel',costfunc,r2,power
      
      
      return
      end

      

      
